### ISSUE TITLE
(Blessed) JSON serializable format

### ISSUE THREAD

**Issue by jakirkham** [numpy/numpy#12481]
It would be really useful for the community at large if NumPy devs proposed a way to make an array JSON serializable and provided builtin functions for performing this serialization/deserialization.

Have taken a quick look at what is out there and found several implementations (a few of these are listed below). Interestingly some people solve this problem by [just using Pandas]( https://stackoverflow.com/a/44752209 ), which only really works for 1-D or 2-D data. Some libraries include their own custom JSON encoders. Other libraries solve this by either providing a grab-bag of JSON solutions including something for NumPy or rolling their own.

While many of the discussions about encoding NumPy arrays as JSON often site that this will not be as performant as binary storage, there are still a large number of implementations for storing NumPy arrays in JSON and people seeking them. Meaning that the use cases driving this serialization are not as much about compact storage, but revolve more around other things like cross-language support, web-friendliness, portability, readability, common tooling support, etc..

Most implementations seem to go the obvious route of turning a NumPy array into a nested list. Though a few base64 encode the entire array. However they vary somewhat on how they pack in the NumPy metadata. Usually this is an dictionary that includes both the metadata and the data as key-value pairs. Some of them include the shape while others do not. Similarly some of them specify whether the data is C ordered. All of them seem to specify dtype in a string. Also they may vary on how they handle non-JSON friendly types (e.g. datetime/timedelta, object, etc.). This data may be coerced to other JSON types (e.g. `None` to `null`) or they may be handled as base64 encoded strings.

Anyways it would be good to have some discussion about MVP functionality that NumPy can provide that gets the ball rolling around consolidating how we talk about NumPy arrays in JSON in the most basic cases. This should provide good context for feedback from the larger community on how to handle various other cases.

Implementations:

ref: http://docs.astropy.org/en/stable/api/astropy.utils.misc.JsonCustomEncoder.html
ref: https://bokeh.pydata.org/en/latest/docs/reference/core/json_encoder.html
ref: https://pythonhosted.org/monty/monty.html#monty.json.MontyEncoder
ref: https://json-tricks.readthedocs.io/en/latest/#numpy-arrays
ref: https://github.com/jsonpickle/jsonpickle#numpy-support

ref: ref: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_json.html

Issues/Questions/Discussions/Blogs:

ref: https://bugs.python.org/issue18303
ref: https://discuss.mxnet.io/t/make-ndarray-json-serializable/1627/2
ref: http://acroz.github.io/2017/11/03/data-science-apis-flask/#jsonifying-numpy-values
ref: https://stackoverflow.com/questions/26646362/numpy-array-is-not-json-serializable
ref: https://stackoverflow.com/questions/43346300/convert-numpy-nd-array-to-json
ref: https://stackoverflow.com/questions/48310067/making-numpy-arrays-json-serializable

**Comment by shoyer** [comment]
I agree that a "multi-dimensional array JSON standard" could be a useful thing, but I'm not sure it makes to develop it as part of NumPy -- this is a broader, cross language discussion. Once such a standard exists, it would be relatively straightforward to integrate it into NumPy. 

Possibly it's worth looking at the example of GeoJSON, which was standardized by the IETF:
http://geojson.org/

**Comment by naught101** [comment]
Very strong support for this. Love the title. Moving small-ish datasets around in a nice way is very useful, and not having to manually write a JSON serialiser or manually dump arrays to lists whenever they appear in dictionaries would be very nice.

GeoJSON is an extremely minimal format, and not really relevant to multi-dimensional data. All it really recommends is a standard way to represent geographical features using particular nested list formats. Not that useful if my 3 dimensions are not geographically related.

It would be really nice to have some way of optionally outputing a simple nested-list format, as well as a more complex dictionary format from which an exact array might be rebuilt. Possibly with something along the lines of `np.set_printoptions`?

**Comment by eric-wieser** [comment]
> It would be really nice to have some way of optionally outputing a simple nested-list format

For simple dtypes, you can get that today with `json.dumps(arr.tolist())`

**Comment by naught101** [comment]
@eric-wieser yes, I know, but it would be good to be able to do this automatically. For instance, I working with an analysis API that uses numpy in many places for efficiency and sanity, but ultimately returns results as deeply nested dictionaries with metadata. The numpy arrays are usually a few layers deep in the dictionary at the end, when it needs to be dumped to JSON, and it's not practical or fun to traverse the dictionary looking for them, nor is it clean or always easy to use `tolist()` where the arrays are added to the dictionary, because sometimes the (partial) dictionaries are used in other stages internally, where it still makes sense to keep things as arrays. At the moment, the code has many if statements checking for these conditions, and it would be nicer if I could just tell numpy to automatically dump to nested lists if it was passed to `json.toJSON`

**Comment by jakirkham** [comment]
I think Stephan's point is GeoJSON is something that could be a starting point to build a cross language N-D array JSON standard that NumPy could use, which makes a lot of sense to me at least. Not so much that GeoJSON should be that thing. Though feel free to correct me if I'm misunderstanding you Stephan.

This would still be useful for Zarr, which is why this came up for us. Today we make a best effort of storing NumPy-based metadata in JSON. Though there are developers of Zarr in other languages that don't have NumPy, who are basically reimplementing the same thing. Having some standard to fill this gap would be welcome. I suppose Zarr can continue to fill this gap on its own

A few things we looked at were [jsonpickle]( http://jsonpickle.github.io/ ) and [json-tricks]( https://json-tricks.readthedocs.io/en/latest/ ). Though these are Python specific, so likely won't have much utility for C, C++, and Java developers for example. Also worth noting is [msgpack-numpy]( https://github.com/lebedov/msgpack-numpy ) though that is obviously MsgPack and not JSON

**Comment by loco-philippe** [comment]
A typed JSON format is defined and applicable also for tabular and multi-dimensional data (see [Internet-Draft JSON-NTV](https://www.ietf.org/archive/id/draft-thomy-json-ntv-02.html) and [Internet-Draft NTV-TAB)](https://www.ietf.org/archive/id/draft-thomy-ntv-tab-00.html). 

This format is implemented for pandas data (see [NTV-pandas](https://github.com/loco-philippe/ntv-pandas/blob/main/README.md) package available in the {pandas ecosystem](https://pandas.pydata.org/community/ecosystem.html) and [PDEP12 specification](https://pandas.pydata.org/pdeps/0012-compact-and-reversible-JSON-interface.html)) and allows a reversible interface (round-trip) for many datatypes.

I could implement an equivalent functionality for Numpy if there is an expectation in this direction?

Would this meet the needs ?

Note: Would it be helpful to write a proposal on the subject?

**Comment by charris** [comment]
Might start a conversation on the mailing list. Can the serialization also store units? That might be useful for astropy, although astronomers have their own interchange formats. @mhvk Any interest?

**Comment by mhvk** [comment]
My general sense is that the moment there is a standard, it is worth supporting it! Things like units are then a question of how to support metadata. I didn't immediately understand this from the descriptions how one would work this in, but the pandas document suggested one would be able to have "composite types", so perhaps something like `Quantity::m::float32` would be possible?

**Comment by loco-philippe** [comment]
@charris @mhvk 

Thank you for these quick and positive responses ! 

I'm going to take some time (this week) to prepare a proposal (simple draft !) that I will share in a conversation on the mailing list.

In this proposal, I will look at the following points:
- Numpy structure
- client structures (e.g. astropy)
- units (e.g. units / Quantities as defined in astropy)
- interchange format for astropy (is ECSV a good reference for that?)

**Comment by mhvk** [comment]
Indeed, `ECSV` is our "solution" to passing around tabular data, basically CSV with a yaml-based header.

**Comment by loco-philippe** [comment]
I start a conversation on the mailing list (waiting moderator approval).

The first draft of the proposal is [this jupyter notebook](https://nbviewer.org/github/loco-philippe/Environmental-Sensing/blob/main/python/Tests/numpy_tests.ipynb)

Note : @mhvk, the Notebook includes a proposal to support units and quantities

**Comment by charris** [comment]
Hmm, I haven't seen any notification of the email. @loco-philippe Where did you send it, did you subscribe?

**Comment by loco-philippe** [comment]
@charris 

I received only the mail below (if there is a problem, i can resend it).

```
From: <numpy-discussion-bounces@python.org>
Date: Tue, Feb 20, 2024 at 12:24 AM
Subject: Your message to numpy-discussion@python.org awaits moderator approval
To: <philippe@loco-labs.io>


Your mail to '[numpy-discussion@python.org](mailto:numpy-discussion@python.org)' with the subject

    JSON format for multi-dimensional data

Is being held until the list moderator can review it for approval.

The message is being held because:

    The message comes from a moderated member

Either the message will get posted to the list, or you will receive
notification of the moderator's decision.
```

**Comment by mattip** [comment]
I tried to approve the mail but it seems the [mailman3 server is offline](https://status.python.org/). Usually it takes up to a day for a mail from new subscribers to be approved.

**Comment by loco-philippe** [comment]
@mattip. Thank you, the mail is now available in the mailing list !

I also completed the proposal formalized in the linked notebook.
It will be completed again in a few days to include the representation of datatypes.

**Comment by loco-philippe** [comment]
The representation of datatypes is now included in the [notebook](https://nbviewer.org/github/loco-philippe/Environmental-Sensing/blob/main/python/Tests/numpy_tests.ipynb).
I will now look at the Xarray part which I will then integrate.

@mhvk, @charris, do you have any remarks or comments about this proposal?

**Comment by mhvk** [comment]
@loco-philippe - I had only a quick look (and sadly do not really have time for more), and felt the proposal looked very reasonable. I think the discussion on the mailing list made clear it is not likely to become part of numpy, though, with the suggestion to have a separate package instead. From the astropy perspective, I think we'd be quite happy to support an external package, including having tests that tables round-trip, etc.

**Comment by loco-philippe** [comment]
@mhvk - I agree with this position of having a separate package (similar solution to what was done for tabular data).

I'm starting now with the goal of having a JSON format that integrates the following points:
- simple multidimensional data (e.g. ndarray Numpy)
- labeled multidimensional data (e.g. DataArray/Dataset Xarray and scipp, NDData astropy) with coordinates, masks, uncertainty and units
- data with numerous types (e.g. Numpy dtype, NTV type, units)
- lossless format (round trip with the main tools e.g. Numpy, Xarray, scipp, astropy)
- NTV structure

The package will contain the conversion functions to the tools indicated above (Numpy first, the others later).

**Comment by loco-philippe** [comment]
@mhvk 

As noted in the previous comment, I created a separate package dedicated to a neutral format for describing and sharing multidimensional data (see [jupyter notebook](https://nbviewer.org/github/loco-philippe/ntv-numpy/blob/main/example/example_ntv_numpy.ipynb), [github repository](https://github.com/loco-philippe/ntv-numpy/blob/main/README.md), [PyPI package](https://pypi.org/project/ntv-numpy/) ).

This provides reversible interfaces (lossless round-trip) with the JSON format and with the following tools:
- Xarray (Dataset, DataArray)
- Astropy (NDData)
- scipp (Dataset)

For example, we can losslessly convert a NDData into an Xarray Dataset or convert it to JSON format.

It also handles easily exchangeable lightweight structures (only metadata that points to URIs to access data stored in independent environments).
Data typing is based on the semantic types defined by the [NTV](https://www.ietf.org/archive/id/draft-thomy-json-ntv-02.html) format..


This package is built based on `numpy.ndarray`.

A second version will integrate tabular representations (integration of the [NTV-TAB](https://www.ietf.org/archive/id/draft-thomy-ntv-tab-00.html) format and the [NTV-pandas](https://github.com/loco-philippe/ntv-pandas) package) and associated interfaces (e.g. pandas).

This first version (alpha) will be completed based on the use cases that will be expressed.

Thank you in advance for your feedback and use cases (github [issues](https://github.com/loco-philippe/ntv-numpy/issues) and [discussions](https://github.com/loco-philippe/ntv-numpy/discussions) are enabled)!

Note: This proposal is also shared with the relevant tools (issue)

**Comment by nstarman** [comment]
This is really cool!

> A second version will integrate tabular representations (integration of the [NTV-TAB](https://www.ietf.org/archive/id/draft-thomy-ntv-tab-00.html) format and the [NTV-pandas](https://github.com/loco-philippe/ntv-pandas) package) and associated interfaces (e.g. pandas).

Just to point out, the tabular interoperability will need to play well with https://data-apis.org/dataframe-api/draft/.

**Comment by loco-philippe** [comment]
@nstarman 

Thank you Nathaniel.

The second version will be uploaded in PyPI ([ntv_numpy](https://pypi.org/project/ntv-numpy/) and [ntv_pandas](https://pypi.org/project/ntv-pandas/))and now we have a lossless interface between multidimensional tools and Pandas (see [example](https://nbviewer.org/github/loco-philippe/ntv-pandas/blob/main/example/example_multidimensional.ipynb)).

The direction pandas -> xarray / scipp is interesting.

A third version will merge JSON tabular format and JSON multidimensional format. 

I had an exchange (issue [#786](https://github.com/data-apis/array-api/issues/786)) with data-api but the answer was "interesting, but out of scope for this project".

**Comment by jakirkham** [comment]
Wow thanks Philippe! 🙏

Am impressed with all of the progress you have made so far 👏

<hr>

Think the next steps will be building awareness around the work you have already done. Some thoughts along these lines:

1. Have you consider writing up a blogpost?
2. Have you reached out to other projects to see if this may fit their needs?
3. Are you going to any conferences this year? Maybe worth doing a talk or a lightning talk?
    1. [PyCon US]( https://us.pycon.org/2024/ ) is next week (so probably too soon unless you are already going)
    2. [SciPy]( https://www.scipy2024.scipy.org/ ) is in early July (maybe an option)
    3. [EuroPython]( https://ep2024.europython.eu ) also early July (so probably need to pick between this and SciPy)
    4. [EuroSciPy]( https://www.euroscipy.org/2024 ) is in late August (and [they have an open CfP]( https://www.euroscipy.org/2024/program.html ))
    5. Others I'm likely not thinking of...?

These other discussion points around standardization, integration, etc. tend to go better when there is broader community buy-in, which the suggestions above might help galvanize 🙂

**Comment by loco-philippe** [comment]
Thank you for this encouragement and support!

Indeed, you are right: it is important to publicize the work done and obtain the support of a community.

For Standardization, I tested the IETF: A written [Internet-Draft](https://www.ietf.org/archive/id/draft-thomy-json-ntv-02.html) was appreciated and the subject was presented at IETF118 but to transform it into an RFC you have to obtain the support of one of the working groups , which cannot be achieved without the support of a community...

For projects, Xarray is interested in pandas_to_xarray conversion and the NTV-pandas package will be integrated into the user guide ([PR #9020](https://github.com/pydata/xarray/pull/9020)) . I will then be able to complete the existing referencing on the pandas side.

For blog posts, yes, it's important and I want to work on it (do you have an idea for a platform to use for this?).

Thank you also for the list of conferences. I will follow your advice and answer the EuroSciPy CfP (the conferences are difficult for me because my level of English is not good and I am not at all comfortable...).

I am also looking to exchange with a mathematician interested in the analysis of tabular structures to share the work done ([see presentation](https://github.com/loco-philippe/tab-analysis/blob/main/docs/tabular_analysis.pdf)) but so far I haven't found anyone.
